components/nonmem/src/signal_wrapper.rs

```diff
@@ -0,0 +1,164 @@
+use signal_hook::consts::{SIGHUP, SIGINT, SIGTERM};
+#[cfg(unix)]
+use std::sync::Arc;
+#[cfg(unix)]
+use std::sync::atomic::{AtomicBool, Ordering};
+#[cfg(unix)]
+use std::thread;
+#[cfg(unix)]
+use std::time::Duration;
+
+pub const TERMINATION_FILENAME: &str = "pharos_terminated.json";
+
+/// Read from a source with a timeout to prevent hanging indefinitely
+#[cfg(unix)]
+fn read_with_timeout(mut reader: impl Read + Send + 'static, timeout: Duration) -> Result<Vec<u8>> {
+    use std::sync::mpsc;
+
+    let (tx, rx) = mpsc::channel();
+
+    thread::spawn(move || {
+        let mut buffer = Vec::new();
+        match reader.read_to_end(&mut buffer) {
+            Ok(_) => {
+                let _ = tx.send(Ok(buffer));
+            }
+            Err(e) => {
+                let _ = tx.send(Err(e));
+            }
+        }
+    });
+
+    match rx.recv_timeout(timeout) {
+        Ok(Ok(buffer)) => Ok(buffer),
+        Ok(Err(e)) => anyhow::bail!("Reading failed: {}", e),
+        Err(_) => {
+            log::warn!(
+                "Reading timed out after {:?}, returning empty output",
+                timeout
+            );
+            Ok(Vec::new()) // Return empty on timeout rather than error
+        }
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Termination {
+    pub signal: String,
+    pub timestamp: String,
+    pub reason: String,
+}
+
+impl Display for Termination {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.write_str(
+            format!(
+                "Terminated by {} ({}) at {}",
+                self.signal, self.reason, self.timestamp
+            )
+            .as_str(),
+        )
+    }
+}
+
+/// Execute command with signal handling - writes termination file if killed
+#[cfg(unix)]
+pub fn execute_with_termination_handling(
+    mut command: Command,
+    recv: impl Read + Send + 'static,
+    output_dir: &Path,
+) -> Result<(std::process::ExitStatus, Vec<u8>)> {
+    // Set up safe flag-based signal handling
+    let sigint_received = Arc::new(AtomicBool::new(false));
+    let sigterm_received = Arc::new(AtomicBool::new(false));
+    let sighup_received = Arc::new(AtomicBool::new(false));
+
+    signal_hook::flag::register(SIGINT, Arc::clone(&sigint_received))?;
+    signal_hook::flag::register(SIGTERM, Arc::clone(&sigterm_received))?;
+    signal_hook::flag::register(SIGHUP, Arc::clone(&sighup_received))?;
+
+    log::info!("Signal handlers registered for SIGINT, SIGTERM, SIGHUP");
+
+    // Spawn child process
+    let mut child = command.spawn()?;
+
+    // Simple main loop: check for signals and child completion
+    loop {
+        // Check for signals first (this makes Ctrl+C responsive)
+        if sigint_received.load(Ordering::Relaxed) {
+            log::info!("SIGINT detected, terminating process");
+            let _ = child.kill(); // Kill child process first
+            write_termination_file(output_dir, "SIGINT", "User interruption (Ctrl+C)")?;
```

Comment:
if any of these write_termination_file ops fail, we'd never exit with the proper exit code, would it be better to swallow this error, printing an error message at least but then exiting